validator/server_api/validation_client.go

```diff
@@ -0,0 +1,214 @@
+	c.LaunchThread(func(ctx context.Context) {
+		input := ValidationInputToJson(entry)
+		var res validator.GoGlobalState
+		err := c.client.CallContext(ctx, &res, Namespace+"_validate", input, moduleRoot)
+		valrun.ConsumeResult(res, err)
+	})
+	return valrun
+}
+
+func (c *ValidationClient) Start(ctx_in context.Context) error {
+	c.StopWaiter.Start(ctx_in, c)
+	ctx := c.GetContext()
+	var client *rpc.Client
+	var err error
+	if len(c.jwtSecret) == 0 {
+		client, err = rpc.DialWebsocket(ctx, c.url, "")
+	} else {
+		client, err = rpc.DialWebsocketJWT(ctx, c.url, "", c.jwtSecret)
+	}
+	if err != nil {
+		return err
+	}
+	var name string
+	err = client.CallContext(ctx, &name, Namespace+"_name")
+	if err != nil {
+		return err
+	}
+	if len(name) == 0 {
+		return errors.New("couldn't read name from server")
+	}
+	c.client = client
+	c.name = name + " on " + c.url
+	return nil
+}
+
+func (c *ValidationClient) Stop() {
+	c.StopWaiter.StopOnly()
+	if c.client != nil {
+		c.client.Close()
+	}
+}
+
+func (c *ValidationClient) Name() string {
+	if c.Started() {
+		return c.name
+	}
+	return "(not started) on " + c.url
+}
+
+func (c *ValidationClient) Room() int {
+	var res int
+	err := c.client.CallContext(c.GetContext(), &res, Namespace+"_room")
+	if err != nil {
+		log.Error("error contacting validation server", "name", c.name, "err", err)
+		return 0
+	}
+	return res
+}
+
+type ExecutionClient struct {
+	ValidationClient
+}
+
+func NewExecutionClient(url string, jwtSecret []byte) *ExecutionClient {
+	return &ExecutionClient{
+		ValidationClient: *NewValidationClient(url, jwtSecret),
+	}
+}
+
+func (c *ExecutionClient) CreateExecutionRun(wasmModuleRoot common.Hash, input *validator.ValidationInput) (validator.ExecutionRun, error) {
+	var res uint64
+	err := c.client.CallContext(c.GetContext(), &res, Namespace+"_createExecutionRun", wasmModuleRoot, ValidationInputToJson(input))
+	if err != nil {
+		return nil, err
+	}
+	return &ExecutionClientRun{c, res}, nil
+}
+
+type ExecutionClientRun struct {
+	client *ExecutionClient
+	id     uint64
+}
+
+func (c *ExecutionClient) LatestWasmModuleRoot() (common.Hash, error) {
+	var res common.Hash
+	err := c.client.CallContext(c.GetContext(), &res, Namespace+"_latestWasmModuleRoot")
+	return res, err
+}
+
+func (c *ExecutionClient) WriteToFile(input *validator.ValidationInput, expOut validator.GoGlobalState, moduleRoot common.Hash) error {
+	jsonInput := ValidationInputToJson(input)
+	err := c.client.CallContext(c.GetContext(), nil, Namespace+"_writeToFile", jsonInput, expOut, moduleRoot)
+	return err
+}
+
+type ExecutionClientStep struct {
+	containers.Promise[validator.MachineStepResult]
+	cancel func()
+}
+
+func (r *ExecutionClientRun) GetStepAt(pos uint64) validator.MachineStep {
+	ctx, cancel := context.WithCancel(r.client.GetContext())
+	step := &ExecutionClientStep{
+		Promise: containers.NewPromise[validator.MachineStepResult](),
+		cancel:  cancel,
+	}
+	go func() {
```

Comment:
Should we use `r.client.LaunchUntrackedThread` here (and for all of the other `go func()`s here) for clarity?